### ISSUE TITLE
Notification email has external FROM address.

### ISSUE THREAD

**Issue by yelizariev** [odoo/odoo#5864]
**Mail server reject notification emails**

Impacted versions:

 - 8.0. Not sure about master, but it seems that there are no changes about this issue at mail_mail.py file.

Steps to reproduce:

 1. set up outgoing mail server via https://pdd.yandex.ru/
 2. set alias for admin: admin@DOMAIN, set email for admin, e.g. ADMIN@gmail.com
 3. activate notification for new emails for admin user
 5. sent external email (e.g. from USER@CUSTOMER) to admin@DOMAIN

Current behavior:

Odoo trying to send notification email to admin user with parameters as follows:
FROM: USER@CUSTOMER
TO: ADMIN@gmail.com

Yandex mail service reject email with error:
SMTPSenderRefused: 553
5.7.1 Sender address rejected: not owned by auth user.

Expected behavior:

FROM value should be replaced, e.g. like this:

FROM: catchall@DOMAIN
TO: ADMIN@gmail.com

There is no such problem when gmail smtp server is used as outgoing mail server at odoo, because gmail replace wrong FROM field with a gmail account, e.g.
FROM: ivan@gmail.com
TO: ADMIN@gmail.com

Some mail service can send email with wrong FROM field, but recepient will get warning that email was sent from another server (such behaviour was in notifcation from odoo forum https://www.odoo.com/forum/help-1 )

possible solution can be found here: https://github.com/yelizariev/addons-yelizariev/blob/8.0/mail_fix_553/mail_fix_553.py

    class mail_mail(osv.Model):
        _inherit = "mail.mail"
    
        def send(self, cr, uid, ids, auto_commit=False, raise_exception=False, context=None):
    
            ...
    
            catchall_alias = self.pool['ir.config_parameter'].get_param(cr, uid, "mail.catchall.alias", context=context)
            catchall_domain = self.pool['ir.config_parameter'].get_param(cr, uid, "mail.catchall.domain", context=context)
    
            correct_email_from = '@%s>?\s*$'%catchall_domain
            default_email_from = '%s@%s' % (catchall_alias, catchall_domain)
    
            ...
    
            for mail in self.browse(cr, SUPERUSER_ID, ids, context=context):
    
                ...
    
                    for email in email_list:
    
                        email_from = mail.email_from
                        if re.search(correct_email_from, email_from) is None:
                            email_from = default_email_from
    
                        msg = ir_mail_server.build_email(
                            email_from=email_from,
                            ...

**Comment by ddfs** [comment]
in codebase: /8.0/nighlty/src/odoo_8.0.20150410.tar.gz, configuring "Alias Domain" is enough.

In Parameters > System Parameters
mail.bounce.alias -> bounce path   [ optional, good to have ]
mail.catchall.alias -> catchall non-tracked emails [ optional, good to have, defaults: postmaster-odoo ]
mail.catchall.domain -> yourdomainname.com [ required ]

**Comment by yelizariev** [comment]
@ddfs, I think we talk about different issues. I have "Alias Domain" setting.

**Comment by mvaled** [comment]
Hi @yelizariev,

This seems related with the DMARC initiative.  Several weeks ago Yahoo.com moved to a reject policy, probably starting a snowball of others email providers changing as well.

I rather keep this issue open and conduct the discussion about solving DMARC here instead of there.

**Comment by mvaled** [comment]
Hi @yelizariev, @odony,

We're actually using an addon that allows to hook into the mail receive/send mechanism of OpenERP.  It is published at https://github.com/merchise-autrement/xopgi.mail_threads.

When we started this addon, the main idea was to extend Odoo's routing mechanism to allow more routing heuristics to be provided.  For instance, we interface with Evaneos (evaneos.com), and its mail system removes the `In-Reply-To` and `References` headers but always uses a unique address to each thread.

With our [MailRouter](https://github.com/merchise-autrement/xopgi.mail_threads#mail-routers) we could fix this issue and selectively (ie. by DB) apply it where it should be.

This fix is been running in our servers for several months now without problems.  

More recently we introduced the concept of [MailTransport](https://github.com/merchise-autrement/xopgi.mail_threads/#mail-transports), that allows to change how to deliver messages.  You may simply change the message's headers, deliver it yourself or chose an outgoing server instead of the default.

The first application of this was to use an external (gmail.com): we wanted to have leads received via a gmail account to be replied with that same account.  A 159 lines transport allows to do that.

At the moment we're testing a VERP implementation that is a simple combination of a router and a transport, plus a transient model (the one bounced messages are routed to) that allows us to attach the bounces to the right thread and be visible to users.

Now we're building a transport will be used to replace the outgoing messages with a DMARC-friendly From header. 

Best regards,
Manuel.

**Comment by mvaled** [comment]
Notice:  Though we deem the xopgi.mail_thread package quite stable we may change its API in the future.  We provide use a X.Y version scheme to easily integrate with OpenEPR's version scheme for addons (e.g: 8.0.2.2).  However, a major change should change the X in our addon.

**Comment by michaelkarrer81** [comment]
+1 i think this "rewirte / fix" of the outgoing mail header will be needed in the future even if i will miss the possibility to search and see the original sender in my external mail box.

What about sending it as a forwarded e-mail? - In fact this is what is is...

**Comment by mvaled** [comment]
Hi @michaelkarrer81,

There's a raw_email where we "keep" the original headers.

**Comment by jdegoy** [comment]
@ yelizariev. If I understand correctly your fix will address the following situation;
1) external@customer.net sends an e-mail to myalias@mymessagingdomain.odoo.com
2) Odoo has Alias Domain set to mymessagingdomain.odoo.com and myalias@mymessagingdomain.odoo.com is being read by Odoo (probably via one generic mailbox which has myalias@mymessagingdomain.odoo.com as an e-mail alias)
3) user is defined with myalias@mymessagingdomain.odoo.com as messageing alias and forwarding is set to All
4) user e-mail (on partner) is therealme@gmail.com
5) with your fix odoo will forward the message from external@customer.net to therrealme@gmail.com using the address specified in mail.catchall.alias_from as FROM instead of external@customer.net.
Nice but personnaly I have not had the restrictions with the smtp server I'm using.

Now, here is what I find another problem on the FROM matter for e-mails.
When I set a messaging alias on a user I expect it to be used as FROM for any e-mail the user in question will generate out of Odoo (for example send a Quotation via e-mail). And this without having to tweak every e-mail template.
In fact, when one removes the e-mail address from a user (via Preferences) and the user has a messaging alias set then Odoo will use the alias as FROM. I'd like it to use the alias as soon as it's set. This still allows for alias and actual e-mail to be equal to each other but gives you the option of hiding your therrealme@gmail.com address if you like by using a messaging alias different from your real e-mail address while still being notified in your real mailbox thru the forwarding mechanism.
Perhaps I should create another issue for this...

**Comment by rafaelbn** [comment]
cc @pedrobaeza @antespi @Yajo @sergio-incaser @carlos-incaser

**Comment by rafaelbn** [comment]
@yelizariev it's much more effective if you make a PR to Odoo fixing it.
Also you can make a PR to https://github.com/OCA/OCB fixing the problem and have it solve it
:+1:

**Comment by odony** [comment]
Closing as a duplicate of #3347, which IMO hold much more info and discussion about the real constraints behind this.

**Comment by dreispt** [comment]
@odony #3347 is closed. Shouldn't it be be reopened?

**Comment by odony** [comment]
@dreispt the issue in #3347 was closed by the OP after #3393 was merged. But the point of marking as duplicate is that the discussion at the beginning of #3347 explains why Odoo works this way, the constraints behind this, why there is no bug here, and the use of the `Sender` header.
Users should also be aware of the limitations of their outgoing SMTP providers. To use Odoo in multi-user environments you need a true SMTP relay that accepts arbitrary From values, exactly as if you were running a mailing-list software (which you actually are). Typically GMail will not work well, as it is a personal SMTP, not a corporate one.

**Comment by rafaelbn** [comment]
@odony this has been never solved or improved. Take a look here with Fabien answer.

https://twitter.com/Chris00393885/status/1231560354515341313

DMARC is implemented in many companies, each month more and more

"Let us configure Odoo to force a FROM address (envelope & header). Let's say I use notifications@mydomain. Then, instead of Odoo trying to send from "Mitchell Admin" <admin@yourcompany.example.com>, it should send from "Mitchell Admin" <notifications@mydomain>, just like Github"

https://twitter.com/__yajo/status/1231856438970114048

**Comment by rafaelbn** [comment]
@Yajo don't loose this please.
